chainstate/test-suite/src/tests/fungible_tokens_v1.rs

```diff
@@ -0,0 +1,2141 @@
+#[case(Seed::from_entropy())]
+fn token_issuance_output_cannot_be_spent(#[case] seed: Seed) {
+    utils::concurrency::model(move || {
+        let mut rng = make_seedable_rng(seed);
+        let mut tf = TestFramework::builder(&mut rng).build();
+
+        let issuance = make_issuance(&mut rng, TokenTotalSupply::Unlimited);
+        let tx = TransactionBuilder::new()
+            .add_input(
+                TxInput::from_utxo(tf.genesis().get_id().into(), 0),
+                InputWitness::NoSignature(None),
+            )
+            .add_output(TxOutput::TokensOp(TokenOutput::IssueFungibleToken(
+                Box::new(issuance.clone()),
+            )))
+            .add_output(TxOutput::Burn(OutputValue::Coin(
+                tf.chainstate.get_chain_config().token_min_issuance_fee(),
+            )))
+            .build();
+        let tx_id = tx.transaction().get_id();
+        tf.make_block_builder().add_transaction(tx).build_and_process().unwrap();
+
+        let result = tf
+            .make_block_builder()
+            .add_transaction(
+                TransactionBuilder::new()
+                    .add_input(
+                        TxInput::from_utxo(tx_id.into(), 0),
+                        InputWitness::NoSignature(None),
+                    )
+                    .add_output(TxOutput::Transfer(
+                        OutputValue::Coin(Amount::ZERO),
+                        Destination::AnyoneCanSpend,
+                    ))
+                    .build(),
+            )
+            .build_and_process();
+
+        assert_eq!(
+            result.unwrap_err(),
+            ChainstateError::ProcessBlockError(BlockError::StateUpdateFailed(
+                ConnectTransactionError::MissingOutputOrSpent(common::chain::UtxoOutPoint::new(
+                    tx_id.into(),
+                    0
+                ))
+            ))
+        );
+    });
+}
+
+// Issue a token with fixed supply.
+// Mint tokens over total supply, check an error.
+// Mint tokens in supply range, check storage.
+// Unmint more tokens that been minter, check an error.
+// Unmint tokens in minter range, check storage.
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn mint_unmint_fixed_supply(#[case] seed: Seed) {
+    utils::concurrency::model(move || {
+        let mut rng = make_seedable_rng(seed);
+        let mut tf = TestFramework::builder(&mut rng).build();
+        let mut nonce = AccountNonce::new(0);
+
+        let token_min_supply_change_fee =
+            tf.chainstate.get_chain_config().token_min_supply_change_fee();
+
+        let total_supply = Amount::from_atoms(rng.gen_range(2..100_000_000));
+        let (token_id, _, utxo_with_change) =
+            issue_token_from_genesis(&mut rng, &mut tf, TokenTotalSupply::Fixed(total_supply));
+
+        let amount_to_mint = Amount::from_atoms(rng.gen_range(1..total_supply.into_atoms()));
+        let amount_to_mint_over_limit = (total_supply + Amount::from_atoms(1)).unwrap();
```

Comment:
Same point with the other comment. It's also (in addition) interesting to see total_supply minted, then +1 minted, and expect it to fail. Even repeatedly after that, unmint 1, then try to mint 2, then unmint a random N, then attempt to mint N+1, then attempt to mint N again after that, and it should finally work.